### ISSUE TITLE
Verify each downloaded block while erasure decoding

### ISSUE THREAD

**Issue by lpoli** [0chain/gosdk#543]
Current download code simply downloads certain number of blocks and reconstructs using erasure encoding.
The problem arises when some dishonest blobber sends wrong data. This data will be included in data-reconstruction so resultant data can be corrupted.

We need to use `encoder.Verify` method to verify if the data from all the blobbers are authentic. If it is false, send it to some function that can determine faulty blobber.
To verify data, we need to get data from `num_of_data_shards + 1`. If we only get data from `num_of_data_shards` then it will be true regardless.
`num_of_data_shards +1` can verify and find faulty blobber only if there is single faulty blobber.

In case there are more than 1 faulty blobbers, We can download same block from all remaining blobbers and find out all the faulty blobbers if in some case `encoder.Verify` returns true.

Now if it fails with all combination then it is different story. User has lost his data.

In summary, we need to verify encoded data and if could not be verified, we need to find faulty blobber/s.

**Comment by sculptex** [comment]
I thought this was already in place for Merkel tree structure as part of miner challenge process?

An alternative would be a suitable hash of recombined chunk (which would be data*blocksize) which could be generated as uploaded.
(Edit: would only validate data to user, not identify bad blobbers' chunks)
